File: Tether/app/src/main/java/com/tether/parent/utils/StringCipher.java
package com.tether.parent.utils;

import android.util.Base64;

import javax.crypto.Cipher;
import javax.crypto.spec.SecretKeySpec;

/**
 * String encryption utility (XOR + AES).
 * File ini Java (bukan Kotlin) - demonstrasi mixed-language project.
 */
public final class StringCipher {

    private static final String XOR_KEY = "Xf2#kL9$mN4@pQ7&wR1*vY6";
    private static final String AES_KEY = "XixF4m1lyS3cur3";
    private static final SecretKeySpec AES_SPEC =
            new SecretKeySpec(AES_KEY.getBytes(), "AES");

    private StringCipher() {
        // Utility class - prevent instantiation
    }

    public static String xorDecrypt(String encrypted) {
        byte[] bytes = encrypted.getBytes(java.nio.charset.StandardCharsets.UTF_8);
        byte[] keyBytes = XOR_KEY.getBytes(java.nio.charset.StandardCharsets.UTF_8);
        byte[] result = new byte[bytes.length];
        for (int i = 0; i < bytes.length; i++) {
            result[i] = (byte) (bytes[i] ^ keyBytes[i % keyBytes.length]);
        }
        return new String(result, java.nio.charset.StandardCharsets.UTF_8);
    }

    public static String xorEncrypt(String plain) {
        byte[] bytes = plain.getBytes(java.nio.charset.StandardCharsets.UTF_8);
        byte[] keyBytes = XOR_KEY.getBytes(java.nio.charset.StandardCharsets.UTF_8);
        byte[] result = new byte[bytes.length];
        for (int i = 0; i < bytes.length; i++) {
            result[i] = (byte) (bytes[i] ^ keyBytes[i % keyBytes.length]);
        }
        return new String(result, java.nio.charset.StandardCharsets.UTF_8);
    }

    public static String decrypt(String encryptedBase64) {
        try {
            Cipher cipher = Cipher.getInstance("AES/ECB/PKCS5Padding");
            cipher.init(Cipher.DECRYPT_MODE, AES_SPEC);
            byte[] decoded = Base64.decode(encryptedBase64, Base64.NO_WRAP);
            return new String(cipher.doFinal(decoded), java.nio.charset.StandardCharsets.UTF_8);
        } catch (Exception e) {
            return xorDecrypt(encryptedBase64);
        }
    }

    public static String encrypt(String plain) {
        try {
            Cipher cipher = Cipher.getInstance("AES/ECB/PKCS5Padding");
            cipher.init(Cipher.ENCRYPT_MODE, AES_SPEC);
            byte[] encrypted = cipher.doFinal(plain.getBytes(java.nio.charset.StandardCharsets.UTF_8));
            return Base64.encodeToString(encrypted, Base64.NO_WRAP);
        } catch (Exception e) {
            return xorEncrypt(plain);
        }
    }

    public static String decode(String encoded) {
        if (encoded.startsWith("ENC:")) {
            return decrypt(encoded.substring(4));
        }
        return encoded;
    }
}

File: XiXFamily/app/src/main/java/com/xixfamily/parent/utils/StringCipher.kt
package com.xixfamily.parent.utils

import android.util.Base64
import javax.crypto.Cipher
import javax.crypto.spec.SecretKeySpec

object StringCipher {
    
    private val XOR_KEY = "Xf2#kL9\u0024mN4@pQ7&wR1*vY6"
    private const val AES_KEY = "XixF4m1lyS3cur3"
    private val aesSpec = SecretKeySpec(AES_KEY.toByteArray(), "AES")
    
    fun xorDecrypt(encrypted: String): String {
        val bytes = encrypted.toByteArray(Charsets.UTF_8)
        val keyBytes = XOR_KEY.toByteArray(Charsets.UTF_8)
        val result = ByteArray(bytes.size)
        for (i in bytes.indices) {
            result[i] = (bytes[i].toInt() xor keyBytes[i % keyBytes.size].toInt()).toByte()
        }
        return String(result, Charsets.UTF_8)
    }
    
    fun xorEncrypt(plain: String): String {
        val bytes = plain.toByteArray(Charsets.UTF_8)
        val keyBytes = XOR_KEY.toByteArray(Charsets.UTF_8)
        val result = ByteArray(bytes.size)
        for (i in bytes.indices) {
            result[i] = (bytes[i].toInt() xor keyBytes[i % keyBytes.size].toInt()).toByte()
        }
        return String(result, Charsets.UTF_8)
    }
    
    fun decrypt(encryptedBase64: String): String {
        return try {
            val cipher = Cipher.getInstance("AES/ECB/PKCS5Padding")
            cipher.init(Cipher.DECRYPT_MODE, aesSpec)
            val decoded = Base64.decode(encryptedBase64, Base64.NO_WRAP)
            String(cipher.doFinal(decoded), Charsets.UTF_8)
        } catch (e: Exception) {
            xorDecrypt(encryptedBase64)
        }
    }
    
    fun encrypt(plain: String): String {
        return try {
            val cipher = Cipher.getInstance("AES/ECB/PKCS5Padding")
            cipher.init(Cipher.ENCRYPT_MODE, aesSpec)
            val encrypted = cipher.doFinal(plain.toByteArray(Charsets.UTF_8))
            Base64.encodeToString(encrypted, Base64.NO_WRAP)
        } catch (e: Exception) {
            xorEncrypt(plain)
        }
    }
    
    fun decode(encoded: String): String {
        return if (encoded.startsWith("ENC:")) {
            decrypt(encoded.substring(4))
        } else {
            encoded
        }
    }
}
